File: decodeways.cpp
class Solution {
public:
    int numDecodings(string s) {
        
        int n = s.length();
        int output[n + 1];
        
        for (int i = 0; i < n; i++) {
            output[i] = 0;
        }
        output[n] = 1;
        
        if (s[n - 1] != '0') {
            output[n - 1] = 1;
        }
        
        for (int i = n - 2; i > -1; i--) {
            if (s[i] != '0') {
                output[i] += output[i + 1];
            }
            if (s[i] == '1' || (s[i] == '2' && s[i + 1] < '7')) {
                output[i] += output[i + 2];
            }
        }
        
        return output[0];
        
    }
};

File: decodeways.py
# 91. Decode Ways

def num_decodings(self, s: str) -> int:
  
  def is_nonzero(index):
    return s[index] != '0'

  def is_two_digit_number(index):
      return s[index] == '1' or (s[index] == '2' and s[index + 1] < '7')

  n = len(s)
  output = [0 for _ in range(n + 1)]
  output[n] = 1
  if is_nonzero(n - 1):
      output[n - 1] = 1

  for i in range(n - 2, -1, -1):
      if is_nonzero(i):
          output[i] += output[i + 1]
      if is_two_digit_number(i):
          output[i] += output[i + 2]

  return output[0]
